XenAdmin/Core/Telemetry.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Reflection;
+using System.Threading;
+using System.Web.Script.Serialization;
+using XenAdmin.Plugins;
+using XenAPI;
+
+namespace XenAdmin.Telemetry
+{
+    internal struct XenCenterMetadata
+    {
+        public SystemInfo System;
+        public XenCenterSettings Settings;
+        public XenCenterInfrastructure Infrastructure;
+        public List<Plugin> Plugins;
+        public string Now;
+    }
+
+    internal struct SystemInfo
+    {
+        public string Version;
+        public string DotNetVersion;
+        public string Culture;
+        public string OsVersion;
+        public string OsCulture;
+        public string IpAddress;
+    }
+
+    internal struct CFU
+    {
+        public bool AllowXenCenterUpdates;
+        public bool AllowPatchesUpdates;
+        public bool AllowXenServerUpdates;
+    }
+
+    internal struct Proxy
+    {
+        public bool UseProxy;
+        public bool UseIEProxy;
+        public bool BypassProxyForServers;
+        public bool ProxyAuthentication;
+        public string ProxyAuthenticationMethod;
+    }
+
+    internal struct SaveAndRestore
+    {
+        public bool SaveSessionCredentials;
+        public bool RequireMasterPassword;
+    }
+
+    internal struct XenCenterSettings
+    {
+        public CFU CFU;
+        public Proxy Proxy;
+        public SaveAndRestore SaveAndRestore;
+        public string HelpLastUsed;
+    }
+
+    internal struct XenCenterInfrastructure
+    {
+        public int TotalConnections;
+        public int Connected;
+    }
+
+    internal struct Plugin
+    {
+        public string Name;
+        public string Organization;
+        public bool Enabled;
+    }
+
+    public static class XenCenterTelemetry
+    {
+        public static string GenerateMetadata(PluginManager pluginManager)
+        {
+            var metadata = new XenCenterMetadata
+            {
+                System = new SystemInfo
+                {
+                    Version = Assembly.GetExecutingAssembly().GetName().Version.ToString(),
+                    DotNetVersion = Environment.Version.ToString(4),
+                    Culture = Thread.CurrentThread.CurrentUICulture.EnglishName,
+                    OsVersion = Environment.OSVersion.ToString(),
+                    OsCulture = CultureInfo.InstalledUICulture.EnglishName,
+                    IpAddress = GetLocalIPAddress()

```

Comment:
@stephen-turner In the case of null objects, the JavaScriptSerailizer will output `"IpAddress":null`. Is this OK? Otherwise I can change the GetLocalIPAddress() function to return an empty string instead of null.